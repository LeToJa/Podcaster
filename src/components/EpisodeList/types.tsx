import { RSSResponse } from '../../helpers/RSSParser/types'

export interface EpisodeListTypes {
    podcastId: string
    episodes: RSSResponse[] | false
}