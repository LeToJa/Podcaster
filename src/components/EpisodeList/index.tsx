import EpisodeCard from '../EpisodeCard'

import { EpisodeListTypes } from './types'

const EpisodeList = ({ podcastId, episodes }: EpisodeListTypes) => {
    if (episodes === false) return <section className='shadow-xl p-4'>This podcast doesn&apos;t use RSS :(</section>

    return <>
        <figure className='shadow-xl p-4 mb-6'>
            <h4 className='text-xl'>{episodes.length} episodes</h4>
        </figure>
        <figure className='shadow-xl p-4'>
            <table className='w-full table-auto'>
                <thead>
                    <tr>
                        <th>Title</th>
                        <th>Date</th>
                        <th>Duration</th>
                    </tr>
                </thead>
                <tbody>
                    {
                        episodes.length ?
                            episodes.map((episode, key) => <EpisodeCard key={key} podcastId={`${podcastId}`} episodeId={key} title={episode.title} date={episode.isoDate.substring(0, 10)} duration={episode.itunes.duration} />) :
                            <tr>
                                <td colSpan={3}>No episodes :(</td>
                            </tr>
                    }
                </tbody>
            </table>
        </figure>
    </>
}

export default EpisodeList