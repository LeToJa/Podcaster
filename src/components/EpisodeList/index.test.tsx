import { render } from '@testing-library/react'
import { describe, it, expect, vi } from 'vitest'

import EpisodeList from '.'

vi.mock('../EpisodeCard', () => ({
    default: ({ title, date, duration }: { title: string, date: string, duration: string }) => (
        <tr>
            <td>{title}</td>
            <td>{date}</td>
            <td>{duration}</td>
        </tr>
    ),
}))

const mockEpisodes = [
    {
        title: 'Episode 1',
        content: '<p>Episode content in HTML format</p>',
        isoDate: '2023-08-29',
        enclosure: {
            url: 'http://example.com/audio.mp3',
            type: 'audio/mpeg',
        },
        itunes: {
            duration: '30:00'
        }
    },
    {
        title: 'Episode 2',
        content: '<p>Episode content in HTML format</p>',
        isoDate: '2023-08-28',
        enclosure: {
            url: 'http://example.com/audio.mp3',
            type: 'audio/mpeg',
        },
        itunes: {
            duration: '45:00'
        }
    }
]

describe('EpisodeList', () => {
    it('renders correctly with episodes (snapshot)', () => {
        const { container } = render(<EpisodeList podcastId='123' episodes={mockEpisodes} />)

        expect(container).toMatchSnapshot()
    })

    it('renders correctly with no episodes (snapshot)', () => {
        const { container } = render(<EpisodeList podcastId='123' episodes={[]} />)

        expect(container).toMatchSnapshot()
    })

    it('renders correctly with no RSS (snapshot)', () => {
        const { container } = render(<EpisodeList podcastId='123' episodes={false} />)

        expect(container).toMatchSnapshot()
    })

    it('displays the correct number of episodes', () => {
        const { getByText } = render(<EpisodeList podcastId='123' episodes={mockEpisodes} />)

        expect(getByText(/2 episodes/i)).toBeTruthy()
    })

    it('renders "No episodes" when the list is empty', () => {
        const { getByText } = render(<EpisodeList podcastId='123' episodes={[]} />)

        expect(getByText(/No episodes :\(/i)).toBeTruthy()
    })

    it('renders "This Podcast doesnt use RSS :(" when there are no episodes', () => {
        const { getByText } = render(<EpisodeList podcastId='123' episodes={false} />)

        expect(getByText(/This Podcast doesn't use RSS :\(/i)).toBeTruthy()
    })

    it('renders the episode details correctly', () => {
        const { getByText } = render(<EpisodeList podcastId='123' episodes={mockEpisodes} />)

        expect(getByText('Episode 1')).toBeTruthy()
        expect(getByText('2023-08-29')).toBeTruthy()
        expect(getByText('30:00')).toBeTruthy()
        expect(getByText('Episode 2')).toBeTruthy()
        expect(getByText('2023-08-28')).toBeTruthy()
        expect(getByText('45:00')).toBeTruthy()
    })
})
