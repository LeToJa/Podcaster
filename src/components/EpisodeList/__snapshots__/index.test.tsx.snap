// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EpisodeList > renders correctly with episodes (snapshot) 1`] = `
<div>
  <figure
    class="shadow-xl p-4 mb-6"
  >
    <h4
      class="text-xl"
    >
      2
       episodes
    </h4>
  </figure>
  <figure
    class="shadow-xl p-4"
  >
    <table
      class="w-full table-auto"
    >
      <thead>
        <tr>
          <th>
            Title
          </th>
          <th>
            Date
          </th>
          <th>
            Duration
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Episode 1
          </td>
          <td>
            2023-08-29
          </td>
          <td>
            30:00
          </td>
        </tr>
        <tr>
          <td>
            Episode 2
          </td>
          <td>
            2023-08-28
          </td>
          <td>
            45:00
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</div>
`;

exports[`EpisodeList > renders correctly with no RSS (snapshot) 1`] = `
<div>
  <section
    class="shadow-xl p-4"
  >
    This podcast doesn't use RSS :(
  </section>
</div>
`;

exports[`EpisodeList > renders correctly with no episodes (snapshot) 1`] = `
<div>
  <figure
    class="shadow-xl p-4 mb-6"
  >
    <h4
      class="text-xl"
    >
      0
       episodes
    </h4>
  </figure>
  <figure
    class="shadow-xl p-4"
  >
    <table
      class="w-full table-auto"
    >
      <thead>
        <tr>
          <th>
            Title
          </th>
          <th>
            Date
          </th>
          <th>
            Duration
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            colspan="3"
          >
            No episodes :(
          </td>
        </tr>
      </tbody>
    </table>
  </figure>
</div>
`;
